### PR TITLE
Include session getter support

### PR DESCRIPTION
**Pull request by hugopellissari**
As our team tries to migrate our tests to use factory_boy, we found tricky to set the DB sessions during the tests runtime, mocking was tricky and nasty. I [found this issue](https://github.com/FactoryBoy/factory_boy/issues/304), which closely resemble the problem we are currently facing, and decided to give a shot.

The idea is to allow passing a callable returning the SQLAlchemy session in a Meta attribute called `sqlalchemy_session_factory`. This has several use-cases, allowing control of the session on runtime.

This is my first OSS contribution ever, super excited with it. Any feedback is appreciated :) 

Fixes https://github.com/FactoryBoy/factory_boy/issues/304

### CODE REVIEW

**Review by francoisfreitag — COMMENTED** [review]
Thanks for offering a PR, with tests! :tada: 
Looks to be addressing a pain point for some SQLAlchemy users, that’s great.

File: tests/test_alchemy.py
Comment (hugopellissari) [reply tests/test_alchemy.py]: Good idea! Done in [097db48](https://github.com/FactoryBoy/factory_boy/pull/927/commits/097db48aa701b178a88e991d33c74475eb97bcff)

File: factory/alchemy.py
Comment (francoisfreitag) [inline factory/alchemy.py:100]: That won’t work, because the modified metaclass will remain modified for the factory. The options should be checked at import time (probably using a [checker](https://github.com/FactoryBoy/factory_boy/blob/d2447abbbf4da99d9e76909b8657524cf8e1c48a/factory/base.py#L117)).

Example test modification leading to a failure:

```diff
diff --git a/tests/test_alchemy.py b/tests/test_alchemy.py
index 6a3ff62..98ac188 100644
--- a/tests/test_alchemy.py
+++ b/tests/test_alchemy.py
@@ -287,6 +287,7 @@ class SQLAlchemySessionFactoryTestCase(unittest.TestCase):
     def test_create_get_session_from_sqlalchemy_session_factory(self):
         SessionGetterFactory.create()
         self.assertEqual(SessionGetterFactory._meta.sqlalchemy_session, models.session)
+        SessionGetterFactory.create()
 
     def test_create_raise_exception_sqlalchemy_session_factory_not_callable(self):
         message = "Provide either a sqlalchemy_session or a sqlalchemy_session_factory, not both"
```

```
ERROR: test_create_get_session_from_sqlalchemy_session_factory (tests.test_alchemy.SQLAlchemySessionFactoryTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/freitafr/dev/factory_boy/tests/test_alchemy.py", line 290, in test_create_get_session_from_sqlalchemy_session_factory
    SessionGetterFactory.create()
  File "/home/freitafr/dev/factory_boy/factory/base.py", line 528, in create
    return cls._generate(enums.CREATE_STRATEGY, kwargs)
  File "/home/freitafr/dev/factory_boy/factory/alchemy.py", line 52, in _generate
    return super()._generate(strategy, params)
  File "/home/freitafr/dev/factory_boy/factory/base.py", line 465, in _generate
    return step.build()
  File "/home/freitafr/dev/factory_boy/factory/builder.py", line 264, in build
    instance = self.factory_meta.instantiate(
  File "/home/freitafr/dev/factory_boy/factory/base.py", line 317, in instantiate
    return self.factory._create(model, *args, **kwargs)
  File "/home/freitafr/dev/factory_boy/factory/alchemy.py", line 97, in _create
    raise RuntimeError("Provide either a sqlalchemy_session or a sqlalchemy_session_factory, not both")
RuntimeError: Provide either a sqlalchemy_session or a sqlalchemy_session_factory, not both
```

**Comment by aalvrz** [comment]
I just stumbled upon this. Can you provide some examples on this is to be used? Or is this simply just using `sqlalchemy_session_factory` in the Meta class?

**Comment by hugopellissari** [comment]
> I just stumbled upon this. Can you provide some examples on this is to be used? Or is this simply just using `sqlalchemy_session_factory` in the Meta class?

Hey @aalvrz ! Yes, the idea is simply allowing a callable returning a sqlalchemy session to be provided in the Meta attribute instead of a plain session. We were facing several problems using factory boy because of this, as we dynamically create a custom session for our tests. This is a way I found to overcome this problem. Do you have a similar problem currently?

**Comment by aalvrz** [comment]
@hugopellissari Are you trying to create a custom session with transactional behavior for your tests? In order to easily clean the database after each test? This is what I am trying to achieve while at the same time switching to SQLAlchemy 1.4.

I have a feeling this might be just what I need!

**Comment by hugopellissari** [comment]
> @hugopellissari Are you trying to create a custom session with transactional behavior for your tests? In order to easily clean the database after each test? This is what I am trying to achieve while at the same time switching to SQLAlchemy 1.4.
> 
> I have a feeling this might be just what I need!

That's is pretty much what I'm trying to achieve with this! We are already doing this, but now that we are trying to use factory boy, it became tricky to set this custom session. Glad it might help you as well :)

**Comment by aalvrz** [comment]
> > @hugopellissari Are you trying to create a custom session with transactional behavior for your tests? In order to easily clean the database after each test? This is what I am trying to achieve while at the same time switching to SQLAlchemy 1.4.
> > I have a feeling this might be just what I need!
> 
> That's is pretty much what I'm trying to achieve with this! We are already doing this, but now that we are trying to use factory boy, it became tricky to set this custom session. Glad it might help you as well :)

Awesome!

By the way I managed to make that work with Factory Boy and SQLAlchemy 1.3. I wrote a post about it here, you might find it useful: https://aalvarez.me/posts/python-transactional-tests-using-sqlalchemy-pytest-and-factory-boy/

However for SQLAlchemy 1.4 this does not seem to work well because of the new way the session should be used (as a context manager) apparently.

File: factory/alchemy.py
Comment (francoisfreitag) [inline factory/alchemy.py:26]:
```suggestion
        if value and meta.sqlalchemy_session:
```

File: tests/test_alchemy.py
Comment (francoisfreitag) [inline tests/test_alchemy.py:282]: Up to you.
```suggestion
        # Reuse the session obtained from sqlalchemy_session_factory.
```

File: tests/test_alchemy.py
Comment (francoisfreitag) [inline tests/test_alchemy.py:286]: Can tighten the regexp.
```suggestion
        message = r"^Provide either a sqlalchemy_session or a sqlalchemy_session_factory, not both$"
```

File: tests/test_alchemy.py
Comment (francoisfreitag) [inline tests/test_alchemy.py:279]: Can also move the `SessionGetterFactory` declaration to this test. Up to you.

**Review by francoisfreitag — APPROVED** [review]

File: tests/test_alchemy.py
Comment (hugopellissari) [reply tests/test_alchemy.py]: Done in [27f3a77](https://github.com/FactoryBoy/factory_boy/pull/927/commits/27f3a77ee634b7863e20f335a5e413d21db82a2c)

File: tests/test_alchemy.py
Comment (hugopellissari) [reply tests/test_alchemy.py]: Done in [08cc710](https://github.com/FactoryBoy/factory_boy/pull/927/commits/08cc710fca63ba84d55d411a36d564a0f834f946)

File: factory/alchemy.py
Comment (hugopellissari) [reply factory/alchemy.py]: Done in [a89c62c](https://github.com/FactoryBoy/factory_boy/pull/927/commits/a89c62c94b6172129c2d85af96acd9a98f477bfb)

File: tests/test_alchemy.py
Comment (hugopellissari) [reply tests/test_alchemy.py:278]: And finally, done in [3dd6bb3](https://github.com/FactoryBoy/factory_boy/pull/927/commits/3dd6bb3d8d639ac29ab915576c218e4bd357f159) :)

**Comment by francoisfreitag** [comment]
Thanks for the update! The code and tests look good to me.

Before to land this patch, please:
- [ ] squash the commits with a [descriptive commit message](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html)
- [ ] document the attribute in https://factoryboy.readthedocs.io/en/latest/orms.html#sqlalchemy, with an example
- [ ] add an entry to the [changelog](https://github.com/hugopellissari/factory_boy/blob/31e8a41827aea805e0475f4e6aceaa19d84cdaac/docs/changelog.rst)

**Comment by hugopellissari** [comment]
I just updated the docs and changelog, if it looks good, I'll squash the commits right after :)

**Comment by hugopellissari** [comment]
Thank you @francoisfreitag for your patience. Super happy it got merged :)

Any idea when the version 3.3.0 will be released?

**Comment by jeffwidman** [comment]
> This is my first OSS contribution ever, super excited with it. Any feedback is appreciated :)

@hugopellissari congrats! I still remember my first contribution ever, I barely knew how to program and I had no idea how to git force push, so when they asked me to squash, I just created a new PR... 😀  But I kept at it, just like you did here, and I learned so much as a result, even got a job out of it. So I'm thrilled for you on your journey! 🚀

**Comment by francoisfreitag** [comment]
> Any idea when the version 3.3.0 will be released?

That depends on @rbarrois. I think he usually tries to cut releases when he has some bandwidth to monitor issues following the release and squash them.

https://github.com/FactoryBoy/factory_boy/issues/886 is also waiting for a review and seems to be blocking the next release.

**Pull request merged**
